charmcraft/helptexts.py

```diff
@@ -118,3 +133,116 @@ def get_full_help(command_groups, global_options):
     # join all stripped blocks, leaving ONE empty blank line between
     text = '\n\n'.join(block.strip() for block in textblocks) + '\n'
     return text
+
+
+def get_detailed_help(command_groups, global_options):
+    """Produce the text for the detailed help.
+
+    It has the following structure:
+
+    - usage
+    - summary
+    - global options
+    - all commands shown with description, grouped
+    - more help
+    """
+    textblocks = []
+
+    # title
+    textblocks.append(HEADER)
+
+    # summary
+    textblocks.append("Summary:" + GENERAL_SUMMARY)
+
+    # column alignment is dictated by longest common commands names and groups names
+    max_title_len = 0
+    for _, _, commands in command_groups:
+        for cmd in commands:
+            max_title_len = max(len(cmd.name), max_title_len)
+    for title, _ in global_options:
+        max_title_len = max(len(title), max_title_len)
+
+    # leave two spaces after longest title (also considering the ':')
+    max_title_len += 3
+
+    global_lines = ["Global options:"]
+    for title, text in global_options:
+        global_lines.extend(_build_item(title, text, max_title_len))
+    textblocks.append("\n".join(global_lines))
+
+    textblocks.append("Commands can be classified as follows:")
+
+    for _, group_description, commands in command_groups:
+        group_lines = ["{}:".format(group_description)]
+        for cmd in commands:
+            group_lines.extend(_build_item(cmd.name, cmd.help_msg, max_title_len))
+        textblocks.append("\n".join(group_lines))
+
+    textblocks.append(textwrap.dedent("""
+        For more information about a specific command, run 'charmcraft help <command>'.
+    """))
+
+    # join all stripped blocks, leaving ONE empty blank line between
+    text = '\n\n'.join(block.strip() for block in textblocks) + '\n'
+    return text
+
+
+def get_command_help(command_groups, command, arguments):
+    """Produce the text for each command's help.
+
+    It has the following structure:
+
+    - usage
+    - summary
+    - options
+    - other related commands
+    - footer
+    """
```

Comment:
May be worth putting a line for each function parameter, describing its structure. For example "command_group is a list of tuples ..."

Reply:
Thanks, I improved the docstring with the suggestion.